### PR TITLE
Add MemoryMonitor component and actions

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Adds a singleton parallel component responsible for monitoring the memory of each parallel component. Also adds simple actions for each type of parallel component that will either compute or process the size of the local branch of a parallel component.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/MemoryMonitor/Tags.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+namespace Tags {
+/*!
+ * \brief Tag to hold memory usage of the parallel component it's templated on
+ * before it is written to disk.
+ *
+ * \details The types in the unordered_map are as follows:
+ *
+ * std::unordered_map<Time, std::unordered_map<Node/Proc, Memory in MB>>
+ */
+template <typename ParallelComponent>
```
Comment (nilsdeppe) [inline src/Parallel/MemoryMonitor/Tags.hpp]: A design decision that's not clear to me is why you opted to have the component be a template parameter instead of using a map with a string as the key. Then you only need one tag in the DataBox instead of many, which simplifies the initialization action quite a bit. You actually wouldn't even need an initialization action anymore, you could trivially use the generic `AddSimpleTags` one. I think you can do that already.

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp [hunk 2]
```diff
@@ -61,7 +61,7 @@ struct ContributeMemoryData {
                     const ArrayIndex& /*array_index*/, const double time,
                     const int node_or_proc, const double size_in_MB) {
     static_assert(Parallel::is_group_v<ContributingComponent> or
-                  Parallel::is_node_group_v<ContributingComponent>);
+                  Parallel::is_nodegroup_v<ContributingComponent>);
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: move to previous commit?

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp [hunk 3]
```diff
@@ -302,4 +302,47 @@ struct ProcessArray {
         std::make_tuple(time, size_per_node, avg_size));
   }
 };
+
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: I don't really love one file with so many different actions in it. My impression is the actions are all meant to be used in different places (maybe one or two in the same place). I think it's clear add either a subdirectory `MemoryMonitor` with several different files, or maybe something like `ProcessSingletonMemory.hpp`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I have not gone through all the code in detail, I have a couple higher level design questions before I do that. If it's better to talk on a call, just let me know!

File: src/Parallel/MemoryMonitor/Tags.hpp (hunk 1 above)
Comment (knelli2) [reply src/Parallel/MemoryMonitor/Tags.hpp]: When I was writing this, it wasn't clear which components would actually use their tag in the memory monitor and which wouldn't. We always have access to the component we are sizing so it made sense to just template the tag on the component and then grab the tag corresponding to the component when I needed it.

However, with this design, the tags are only really needed for group and nodegroup components so a map of strings probably makes more sense.

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp (hunk 3 above)
Comment (knelli2) [reply src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: Besides tests, all these actions are going to be using in a `MonitorMemory` event I'll add in a future PR. `ContributeMemoryData` won't be called directly by the event but indirectly by the `ProcessGroup` and `ProcessNodegroup` actions.

That being said, this file is a bit crowded with different things and may be confusing. I can split it up into 4 files in a `MemoryMonitor` subdirectory:
- `ContributeMemoryData.hpp`
- `ProcessGroups.hpp` (has both group and nodegroup since they are so similar)
- `ProcessArray.hpp`
- `ProcessSingleton.hpp`

File: src/Parallel/MemoryMonitor/Tags.hpp (hunk 1 above)
Comment (nilsdeppe) [reply src/Parallel/MemoryMonitor/Tags.hpp]: Okay, is it reasonable to make that change now? Probably before doing so it's worth thinking carefully if there will be any downsides. Nothing jumps out at me since we don't really need the compile time info in the tag, it's basically just used for naming things, and so we can just do the naming dynamically

File: src/ParallelAlgorithms/Actions/MonitorMemory.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Actions/MonitorMemory.hpp]: Okay, that sounds good to me if that sounds like the right thing to you. You know all the followup PRs :)

File: src/Parallel/MemoryMonitor/Tags.hpp (hunk 1 above)
Comment (knelli2) [reply src/Parallel/MemoryMonitor/Tags.hpp]: I don't see a downside either. It's basically just making the name a string at runtime vs a template at compile time. Nothing else

**Comment by knelli2** [comment]
@nilsdeppe Made those changes and shuffled a couple things around, but the bulk of the code stayed the same.

File: src/Parallel/MemoryMonitor/Tags.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  return "/MemoryMonitors/" + pretty_type::name<ParallelComponent>();
+}
+
+namespace Tags {
+/*!
+ * \brief Tag to hold memory usage of parallel components before it is written
+ * to disk.
+ *
+ * \details The types in the unordered_map are as follows:
+ *
```
Comment (nilsdeppe) [inline src/Parallel/MemoryMonitor/Tags.hpp:30]: Do you want to make this a code block using triple ticks on the line before and after? Up to you :)

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 5]
```diff
@@ -24,7 +99,48 @@ void test_tags() {
...
+      make_not_null(&runner), ActionTesting::NodeId{0},
+      ActionTesting::LocalCoreId{0}, {});
+
+  std::unordered_map<
+      std::string, std::unordered_map<double, std::unordered_map<int, double>>>
+      empty_holder{};
+
+  using tag = mem_monitor::Tags::MemoryHolder;
+  CHECK(ActionTesting::box_contains<mem_mon_comp, tag>(runner, 0));
+  CHECK(ActionTesting::tag_is_retrievable<mem_mon_comp, tag>(runner, 0));
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: Make this a `REQUIRE` since the rest of the test won't give reasonable results if this doesn't pass?

File: src/Parallel/MemoryMonitor/MemoryMonitor.hpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+struct InitializeMutator {
+  using return_tags = tmpl::list<mem_monitor::Tags::MemoryHolder>;
+  using argument_tags = tmpl::list<>;
+
+  using tag_type = typename mem_monitor::Tags::MemoryHolder::type;
+
+  static void apply(const gsl::not_null<tag_type*> /*holder*/) {}
+};
+}  // namespace mem_monitor::detail
+
```
Comment (nilsdeppe) [inline src/Parallel/MemoryMonitor/MemoryMonitor.hpp:35]: can you document the `mem_monitor` namespace in this file? Not the `detail` one above, that way it'll show up in the dox, as well as everything inside of it. I would argue for putting the component in that namespace too. No reason all components have to be in the global namespace. Then I'd put the `mem_monitor` in the parallel doxygen group, but not the component itself

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+ * - Size on node 2 (MB)
+ * - Proc of max size
+ * - Size on proc of max size (MB)
+ * - Average size per node (MB)
+ *
+ * The dat file will be placed in the `/MemoryMonitors/` group in the reduction
+ * file. The name of the dat file is the `pretty_type::name` of the component.
+ */
+template <typename ContributingComponent>
+struct ContributeMemoryData {
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp:57]: What do you think of changing the name to `ContributeMemoryConsumption` or `MemoryUsed`? Something more specific than `Data`? Feel free to decline or defer to a PR at the end that does a global rename. I realize renames can be rather annoying in the middle of a series of PRs or commits :)

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 8]
```diff
@@ -139,8 +173,215 @@ void test_initialize() {
...
+using metavars = TestMetavarsActions;
+using mem_mon_comp = MockMemoryMonitor<metavars>;
+using obs_writer_comp = MockObserverWriter<metavars>;
+using sing_comp = SingletonParallelComponent<metavars>;
+using group_comp = GroupParallelComponent<metavars>;
+using array_comp = ArrayParallelComponent<metavars>;
+using nodegroup_comp = NodegroupParallelComponent<metavars>;
+
+void setup_runner(
+    gsl::not_null<ActionTesting::MockRuntimeSystem<metavars>*> runner) {
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: `const gsl`

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                    const int node_or_proc, const double size_in_MB) {
+    static_assert(Parallel::is_group_v<ContributingComponent> or
+                  Parallel::is_nodegroup_v<ContributingComponent>);
+
+    using tag = Tags::MemoryHolder;
+    if constexpr (db::tag_is_retrievable_v<tag, db::DataBox<DbTags>>) {
+      db::mutate<tag>(
+          make_not_null(&box),
+          [&cache, &time, &node_or_proc, &size_in_MB](
+              gsl::not_null<std::unordered_map<
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: `const gsl`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 10]
```diff
@@ -139,8 +173,215 @@ void test_initialize() {
...
+  ActionTesting::emplace_singleton_component_and_initialize<mem_mon_comp>(
+      runner, ActionTesting::NodeId{2}, ActionTesting::LocalCoreId{2}, {});
+
+  runner->set_phase(metavars::Phase::Monitor);
+}
+
+template <typename Component>
+void check_output(const std::string& filename, const double time,
+                  const size_t num_nodes, const std::vector<double>& sizes) {
+  h5::H5File<h5::AccessType::ReadOnly> read_file{filename + ".h5"};
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp:192]: we avoid disk IO in these types of tests by having a mock writing action that just records what was meant to be written and then checking that. Disk IO is slow, and just increases likelihood of bugs in general. Eg line 46 of `tests/Unit/Helpers/ParallelAlgorithms/LinearSolver/ResidualMonitorActionsTestHelpers.hpp`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 11]
```diff
@@ -139,8 +173,215 @@ void test_initialize() {
...
+
+  CHECK(data(0, 0) == time);
+  for (size_t i = 0; i < num_nodes; i++) {
+    CHECK(data(0, i + 1) == sizes[i]);
+  }
+  CHECK(data(0, num_columns - 1) == average);
+}  // namespace
+
+template <typename Component>
+void run_group_actions(ActionTesting::MockRuntimeSystem<metavars>& runner,
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: pass runner as `gsl::not_null`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 12]
```diff
@@ -139,8 +173,215 @@ void test_initialize() {
...
+                                                                      0);
+
+  // Before we invoke the actions, the map for this component shouldn't exist
+  // because this is the first time we are calling it.
+  const std::string name = pretty_type::name<Component>();
+  CHECK(mem_holder_tag.count(name) == 0);
+
+  for (size_t i = 0; i < num_branches; i++) {
+    // unordered_map<name, unordered_map<time, unordered_map<node/proc, size>>>
+    // Skip 0, otherwise an access error will occur
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: I'm a bit confused, would that mean you want the check _after_ the action?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 13]
```diff
@@ -139,8 +173,215 @@ void test_initialize() {
...
+
+  const double average =
+      alg::accumulate(sizes, 0.0) / static_cast<double>(num_nodes);
+
+  CHECK(data(0, 0) == time);
+  for (size_t i = 0; i < num_nodes; i++) {
+    CHECK(data(0, i + 1) == sizes[i]);
+  }
+  CHECK(data(0, num_columns - 1) == average);
+}  // namespace
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: isn't this closing a function?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 14]
```diff
@@ -139,8 +173,215 @@ void test_initialize() {
...
+
+        mem_monitor::ContributeMemoryData<group_comp>::template apply<
+            mem_mon_comp>(empty_box, cache, 0, 0.0, 0_st, 0.0);
+      }()),
+      Catch::Contains("Expected the DataBox for the MemoryMonitor"));
+}
+
+// Contribute memory data can only be used with groups or nodegroups
+template <typename Gen>
+void test_contribute_memory_data(gsl::not_null<Gen*> gen) {
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: `const gsl`

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 15]
```diff
@@ -382,6 +426,7 @@ SPECTRE_TEST_CASE("Unit.Parallel.MemoryMonitor", "[Unit][Parallel]") {
   test_tags();
   test_initialize();
   test_wrong_box();
-  test_contribute_memory_data(make_not_null(&gen));
+  test_contribute_memory_data(make_not_null(&gen), false);
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp:487]: maybe add a comment here `// First just test `ContributeMemoryData` (last argument false), then test...`?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 16]
```diff
@@ -421,12 +423,55 @@ void test_contribute_memory_data(gsl::not_null<Gen*> gen,
                                     static_cast<int>(num_procs), time, sizes);
 }
 
+template <typename Gen>
+void test_process_array(gsl::not_null<Gen*> gen) {
```
Comment (nilsdeppe) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: `const gsl`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
overall LGTM, went through the code in detail this time.

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp (hunk 7 above)
Comment (knelli2) [reply src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp:57]: I think I'll wait until all the memory monitor stuff is in just so rebasing goes smoother.

How about `ContributeComponentSize`?

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp (hunk 12 above)
Comment (knelli2) [reply tests/Unit/Parallel/Test_MemoryMonitor.cpp]: So before the first action is invoked on the memory monitor, the map for that time should be empty. And after the last action, the map should also be empty (it gets cleaned up). So if I move the check after the action, I'd have to do something like

```
if (i != num_branches-1) {
  CHECK(...);
}
```

to avoid an access error after the last action. So, it's just a matter of where you want to put the skip.

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp (hunk 7 above)
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp:57]: Yep, both sounds good!

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp (hunk 12 above)
Comment (nilsdeppe) [reply tests/Unit/Parallel/Test_MemoryMonitor.cpp]: Ah okay, could you maybe rephrase the comment to include that detail? That is, `The map is empty both before and after the last call, so we need to put the CHECK inside an if clause regardless`?

**Comment by knelli2** [comment]
@nilsdeppe Rebased on develop and added fixups.

Re: the disk IO comment. I have opened a PR (#3969) to address this generally, and not just for this PR. Since this PR has multiple follow ups that depend on it, if it's ok with you, I'd like to forgo the changes to the test for now so we don't have to wait for #3969. Once #3969 is merged, I can open a follow-up to this PR that changes the test. Does that sound alright?

**Review by nilsdeppe — COMMENTED** [review]
LGTM, and I'm happy leaving the H5 refactor until later!

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@kidder said he would like to review this, so waiting on him

File: src/Parallel/MemoryMonitor/MemoryMonitor.hpp [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+/*!
+ * \ingroup ParallelGroup
+ * Holds the MemoryMonitor parallel component and all actions and tags related
+ * to the memory monitor.
+ */
+namespace mem_monitor {}
+
+namespace mem_monitor {
+namespace detail {
+struct InitializeMutator {
```
Comment (kidder) [inline src/Parallel/MemoryMonitor/MemoryMonitor.hpp:26]: this seems (not your fault) silly...
Not for this PR, but I think `AddSimpleTags` should just take a list of tags which are default initialized (which is already done automatically by the DataBox).  If you want a simple tag not default initialized you should just call an initialization action or `MutateApply` with a mutator.  Then we don't need this detail struct

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+ * file. The name of the dat file is the `pretty_type::name` of the component.
+ */
+template <typename ContributingComponent>
+struct ContributeMemoryData {
+  template <typename ParallelComponent, typename DbTags, typename Metavariables,
+            typename ArrayIndex>
+  static void apply(db::DataBox<DbTags>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/, const double time,
+                    const int node_or_proc, const double size_in_MB) {
```
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: MB -> megabytes

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 19]
```diff
@@ -24,7 +99,48 @@ void test_tags() {
   CHECK(subpath == "/MemoryMonitors/MockMemoryMonitor");
 }
 
+void test_initialize() {
```
Comment (kidder) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: remove this test because (as far as I can tell) it doesn't test anything except the mocking framework...

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+      db::mutate<tag>(
+          make_not_null(&box),
+          [&cache, &time, &node_or_proc, &size_in_MB](
+              const gsl::not_null<std::unordered_map<
+                  std::string,
+                  std::unordered_map<double, std::unordered_map<int, double>>>*>
+                  memory_holder_all) {
+            auto memory_holder_pair = memory_holder_all->try_emplace(
+                pretty_type::name<ContributingComponent>(),
+                std::unordered_map<double, std::unordered_map<int, double>>{});
```
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: you don't need to pass the second argument to get a default constructed object

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+            auto& memory_holder = (*memory_holder_pair.first).second;
+            // If this is the first reduction at this time, create a new
+            // unordered map
+            if (memory_holder.count(time) == 0) {
+              memory_holder.emplace(time, std::unordered_map<int, double>{
+                                              {node_or_proc, size_in_MB}});
```
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: remove the if-else block, and just call `memory_holder.try_emplace(time);` followed by line 87

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+                Parallel::is_group_v<ContributingComponent>;
+
+            const int num_nodes =
+                Parallel::number_of_nodes(*Parallel::local(mem_monitor_proxy));
+            const int num_procs =
+                Parallel::number_of_procs(*Parallel::local(mem_monitor_proxy));
+            const size_t expected_number =
+                static_cast<size_t>(is_group ? num_procs : num_nodes);
+            ASSERT(memory_holder.at(time).size() <= expected_number,
+                   "ContributeMemoryData received more data that it was "
```
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp]: that -> than

File: tests/Unit/Parallel/Test_MemoryMonitor.cpp [hunk 23]
```diff
@@ -139,8 +173,217 @@ void test_initialize() {
...
+  // because this is the first time we are calling it.
+  const std::string name = pretty_type::name<Component>();
+  CHECK(mem_holder_tag.count(name) == 0);
+
+  for (size_t i = 0; i < num_branches; i++) {
+    // unordered_map<name, unordered_map<time, unordered_map<node/proc, size>>>
+    // Putting the CHECK before the action invocation is arbitrary. The map at a
+    // specific time is empty both before the first action is invoked and after
+    // the last action is invoked. Because of this, we need to put the CHECK in
+    // and if statement to avoid an access error, whether the check happens
```
Comment (kidder) [inline tests/Unit/Parallel/Test_MemoryMonitor.cpp]: and -> an

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    static_assert(
+        Parallel::is_nodegroup_v<ParallelComponent>,
+        "ProcessNodegroup can only be run on a Nodegroup parallel component.");
+
+    auto& singleton_proxy =
+        Parallel::get_parallel_component<MemoryMonitor<Metavariables>>(cache);
+
+    auto& nodegroup_proxy =
+        Parallel::get_parallel_component<ParallelComponent>(cache);
+    const int my_node =
```
Comment (kidder) [inline src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp]: isn't this just `array_index` (and thus can't we just have one action that works on both groups and nodegroups)?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp (hunk 24 above)
Comment (knelli2) [reply src/ParallelAlgorithms/Actions/MemoryMonitor/ProcessGroups.hpp]: You're right. `ContribtueMemoryData` expects an `int` for the node/proc so I just `static_assert`ed that the array index was an `int` inside `ProcessGroups`. The only component that doesn't have `int` for the array index is `Array` (and you shouldn't be running this action on an `Array`) so we would probably have been good without it, but I added it just in case.

**Review by kidder — COMMENTED** [review]
you have my permission to squash; @nilsdeppe did you want to look at the fixups?

**Comment by nilsdeppe** [comment]
I'm happy with the fixups and for you to squash and rebase

**Comment by knelli2** [comment]
@kidder @nilsdeppe Squashed and rebased!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
